File: docs/verify-signature.js
// Verifying a hookrelay delivery in Node, with no dependencies.
//
// The signature covers `id.timestamp.rawBody`, so you must hash the bytes you received,
// not a re-serialised object. Express: app.post(path, express.raw({type: '*/*'}), ...)
// and read req.body as a Buffer. JSON.stringify(JSON.parse(body)) is not the same bytes.

const crypto = require('crypto');

const TOLERANCE_SECONDS = 5 * 60;

function verify(secret, headers, rawBody) {
  const id = headers['webhook-id'];
  const timestamp = headers['webhook-timestamp'];
  const received = headers['webhook-signature'];
  if (!id || !timestamp || !received) return false;

  // Reject anything too old to be a live delivery: without this, a signature captured
  // once stays valid forever and can be replayed at any time.
  const age = Math.abs(Math.floor(Date.now() / 1000) - Number(timestamp));
  if (!Number.isFinite(age) || age > TOLERANCE_SECONDS) return false;

  const key = Buffer.from(secret.replace(/^whsec_/, ''), 'base64');
  const expected = 'v1,' + crypto.createHmac('sha256', key)
    .update(`${id}.${timestamp}.${rawBody}`)
    .digest('base64');

  // A rotation sends several signatures, space delimited; any one matching is enough.
  return received.split(' ').some((candidate) => {
    const a = Buffer.from(candidate);
    const b = Buffer.from(expected);
    return a.length === b.length && crypto.timingSafeEqual(a, b);
  });
}

module.exports = { verify };

File: docs/VerifySignature.java
// Verifying a hookrelay delivery in Java, with nothing but the JDK.
//
// The signature covers `id.timestamp.rawBody`, so hash the body exactly as it arrived.
// In Spring, take the body as a String parameter rather than a mapped object, or the
// bytes you verify will not be the bytes that were signed.

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.util.Base64;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public final class VerifySignature {

    private static final long TOLERANCE_SECONDS = 5 * 60;

    public static boolean verify(String secret, String id, String timestamp, String signatureHeader, String rawBody)
            throws Exception {
        if (id == null || timestamp == null || signatureHeader == null) {
            return false;
        }
        // Reject anything too old to be a live delivery. Without this check a signature
        // captured once stays valid forever and can be replayed at any time.
        long age = Math.abs(Instant.now().getEpochSecond() - Long.parseLong(timestamp));
        if (age > TOLERANCE_SECONDS) {
            return false;
        }

        byte[] key = Base64.getDecoder().decode(secret.replaceFirst("^whsec_", ""));
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(key, "HmacSHA256"));
        byte[] digest = mac.doFinal((id + "." + timestamp + "." + rawBody).getBytes(StandardCharsets.UTF_8));
        byte[] expected = ("v1," + Base64.getEncoder().encodeToString(digest)).getBytes(StandardCharsets.UTF_8);

        // A rotation sends several signatures, space delimited; any one matching is enough.
        for (String candidate : signatureHeader.trim().split("\\s+")) {
            if (MessageDigest.isEqual(expected, candidate.getBytes(StandardCharsets.UTF_8))) {
                return true;
            }
        }
        return false;
    }

    private VerifySignature() {}
}
